File: docs/algorithm/simple/code/20. 有效的括号/Code.java
import java.util.ArrayList;
import java.util.List;

class Code {
    public static void main(String[] args) {
        // System.out.println(isValid(args[0]));
        System.out.println((int) '[');
        System.out.println((int) '{');
        System.out.println((int) '(');
        System.out.println((int) ')');
        System.out.println((int) '}');
        System.out.println((int) ']');
        // System.out.println(isValid("[]"));
        // System.out.println(isValid("({[]})"));
        // System.out.println(isValid("]"));
        // System.out.println(isValid("{[]()}"));
        // System.out.println(isValid("{[]}()[][]({[]})"));
    }

    public static boolean isValid2(String s) {
        List<Character> left = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '[' || c == '{' || c == '(') {
                left.add(c);
            } else {
                if (left.size() == 0)
                    return false;
                if (c == ']' && left.get(left.size() - 1) != '[')
                    return false;
                if (c == '}' && left.get(left.size() - 1) != '{')
                    return false;
                if (c == ')' && left.get(left.size() - 1) != '(')
                    return false;
                left.remove(left.size() - 1);
            }
        }
        return left.size() == 0;
    }

    public static boolean isValid(String s) {
        char[] cc = new char[s.length() + 1];
        int top = 1;
        for (char c : s.toCharArray()) {
            if (c == '[' || c == '{' || c == '(') {
                cc[top++] = c;
            } else {
                if (c == ']' && cc[--top] != '[')
                    return false;
                if (c == '}' && cc[--top] != '{')
                    return false;
                if (c == ')' && cc[--top] != '(')
                    return false;
            }
        }
        return top == 1;
    }
}

File: docs/algorithm/simple/code/20. 有效的括号/Code.dart
/// 20. 有效的括号
main() {
  print(isValid("{{}}[()(())]"));
}

bool isValid2(String s) {
  List<String> left = [];
  for (int i = 0; i < s.length; i++) {
    String c = s.substring(i, i + 1);
    if (c == '[' || c == '{' || c == '(') {
      left.add(c);
    } else {
      if (left.length == 0) return false;
      if (c == ']' && left[left.length - 1] != '[') return false;
      if (c == '}' && left[left.length - 1] != '{') return false;
      if (c == ')' && left[left.length - 1] != '(') return false;
      left.removeAt(left.length - 1);
    }
  }
  return left.length == 0;
}

bool isValid(String s) {
  List<String> stack = ['1'];
  int top = 1;
  for (int i = 0; i < s.length; i++) {
    String c = s.substring(i, i + 1);
    print("$c , stack : $stack");
    if (c == '(' || c == '{' || c == '[') {
      stack.add(c);
      top ++;
    } else if (c == ')' && stack[--top] != '(') {
      return false;
    } else if (c == '}' && stack[--top] != '{') {
      return false;
    } else if (c == ']' && stack[--top] != '[') {
      return false;
    }
  }
  return top == 1;
}
